### DISCUSSION TITLE
Introduce SpotBugs Release candidate builds or preview features to improve release quality

### DISCUSSION THREAD

**Discussion by xzel23** [spotbugs/spotbugs#2941] [General]
(Related discussion from two years ago here: https://github.com/spotbugs/spotbugs/discussions/2076)

Hi there, first of all thanks for maintaining SpotBugs.

What do you think about providing and announcing beta and/or release candidate builds? To me it seems that every new version breaks something. I don't mean justified SpotBugs warnings as these should be fixed in the user code. I mean that with every new version, I see lots of false positives in my code. Some maybe cannot be avoided, but often it seems new rules or changes to rules are rolled out before they are ready for prime.

With each new SpotBugs version, I add new exceptions to my exclude file that stay there forever - maybe another nice feature would be to be able to create a report of unnecessary exclusions.

However, I think it would be a good thing to have a means to get better testing by SpotBugs users before a new version is rolled out. One can simply not think about all cases in their tests, and  some testing by the community would certainly improve both release quality and acceptance of SpotBugs.

For example, in the latest release, SpotBugs complains about Singleton issues for a class that has multiple public static factory methods that all return new instances of the class and that is also instantiated via a package private constructor from a builder class in the same package. Someone else reported a similar FP (https://github.com/spotbugs/spotbugs/issues/2934). SpotBugs also complains about instance getters not being synchronised when eager initialisation is used (https://github.com/spotbugs/spotbugs/issues/2932).  Both flag widely used and valid practice as errors. I am sure, many of such false positives could be avoided by having beta or release candidate builds.

What are your thoughts about this?

**Comment by ThrawnCA** [comment]
Beta releases are in principle a good idea to facilitate more testing.

**Comment by hazendaz** [comment]
I'm going to close this for now only because there are really only 3 active contributors, count 4 if we throw in @ThrawnCA as a code reviewer.  Almost everyone else other than occasional comments are not active so we don't have such time to devise this.  While we had a number of issues during a few of our more recent releases do understand we had just went through losing practically every day to day contributor for whatever reasons and two that were involved in other ways were elevated and me personally also was elevated and sort of took over if you will as I needed releases to occur for my only care (the maven plugin).  At any rate, its a great idea, we just don't have the resources to do this at this time.  If you want to take this on though, by all means raise a normal issue and get to coding ;)
